nmmo/core/env.py

```diff
@@ -131,38 +115,22 @@ def action_space(self, agent):
   ############################################################################
   # Core API
 
-  def change_task(self,
-                  new_tasks: List[Union[Tuple[Task, float], Task]],
-                  task_encoding: Optional[Dict[int, np.ndarray]] = None,
-                  embedding_size: int=16,
-                  reset: bool=True,
-                  map_id=None,
-                  seed=None,
-                  options=None):
-    """ Changes the task given to each agent
-
-    Args:
-      new_task: The task to complete and calculate rewards
-      task_encoding: A mapping from eid to encoded task
-      embedding_size: The size of each embedding
-      reset: Resets the environment
-    """
-    self._tasks = [t if isinstance(t, Tuple) else (t,1) for t in new_tasks]
-    self._task_encoding = task_encoding
-    self._task_embedding_size = embedding_size
-    if reset:
-      self.reset(map_id=map_id, seed=seed, options=options)
-
   # TODO: This doesn't conform to the PettingZoo API
   # pylint: disable=arguments-renamed
-  def reset(self, map_id=None, seed=None, options=None):
+  def reset(self, map_id=None, seed=None, options=None,
+            new_tasks: List[Task]=None,
+            task_spec: List[Tuple]=None,
+            teams: Dict[int,List[int]]=None):
     '''OpenAI Gym API reset function
 
     Loads a new game map and returns initial observations
 
     Args:
-        idx: Map index to load. Selects a random map by default
-
+        map_id: Map index to load. Selects a random map by default
+        seed: random seed to use
+        new_tasks: A list of instantiated tasks
+        task_spec: A list of task spec to instantiate inside reset()
```

Comment:
how about just make_new_tasks_fn

that way env doesn't need to know anything about making tasks, or teams

Reply:
Can you please tell me more about it? Let me think about it again after sleep.

However, I also realized that the current `make_team_tasks(teams, task_spec)` doesn't need any info about the env. If one can pass in teams and task_spec, one can also use make_team_tasks to make task instances and pass them in. ... Then yes, the env doesn't need to know anything about making tasks or teams.

Reply:
Oh, I got it, and implemented as follows. But after writing it, I just thought... if one knows it all, why bother passing in these and not the instantiated tasks? Please let me know what you think.
```
    if new_tasks is not None:
      # providing an empty new_tasks [] is also possible
      self.tasks = new_tasks
    elif make_task_fn is not None:
      self.tasks = make_task_fn(**make_task_fn_kwargs)
    else:
      for task in self.tasks:
        task.reset()
```